File: code-python/10_window.py
# # Window Functions

# Copyright © 2010–2018 Cloudera. All rights reserved.
# Not to be reproduced or shared without prior written 
# consent from Cloudera.


# ## Overview

# In this module we demonstrate how to create and apply window functions.


# ## Window Functions

# * Spark SQL supports the following window functions:
#   * `cume_dist`
#   * `dense_rank`
#   * `lag`
#   * `lead`
#   * `ntile`
#   * `percent_rank`
#   * `rank`
#   * `row_number`

# * Aggregate and window functions are applied `over` a window specification

# * A window specification consists of at least one of the following:
#   * Partitioning column
#   * Ordering column
#   * Row specification

# * Windows are experimental as of Spark 2.3.


# ## Setup

# Create a SparkSession:
from pyspark.sql import SparkSession
spark = SparkSession.builder.master("local").appName("window").getOrCreate()

# Read the enhanced ride data from HDFS:
rides = spark.read.parquet("/duocar/joined/")


# ## Example: Cumulative Count and Sum

# Create a simple DataFrame:
df = spark.range(10)
df.show()

# Create a simple window specification:
from pyspark.sql.window import Window
ws = Window.rowsBetween(Window.unboundedPreceding, Window.currentRow)
type(ws)

# Use the window specification to compute cumulative count and sum:
from pyspark.sql.functions import count, sum
df.select("id", count("id").over(ws).alias("cum_cnt"), sum("id").over(ws).alias("cum_sum")).show()

# **Tip:** Examine the default column name to gain additional insight (if you
# are SQL literate):
df.select("id", count("id").over(ws), sum("id").over(ws)).printSchema()


# ## Example: Compute average days between rides for each rider

# Create window specification:
ws = Window.partitionBy("rider_id").orderBy("date_time")

# Use the `lag` function to extract the date and time of the previous ride:
from pyspark.sql.functions import lag
rides2 = rides.withColumn("previous_ride", lag("date_time").over(ws))
rides2.select("rider_id", "date_time", "previous_ride").show(truncate=False)

# **Note:** A rider's first ride does not have a previous ride; therefore, the
# value is set to null.

# Compute the number of days between consecutive rides:
from pyspark.sql.functions import datediff
rides3 = rides2.withColumn("days_between_rides", datediff("date_time", "previous_ride"))
rides3.select("rider_id", "date_time", "previous_ride", "days_between_rides").show(truncate=False)

# Compute the average days between rides for each rider:
from pyspark.sql.functions import count, mean
rides4 = rides3 \
  .groupBy("rider_id") \
  .agg(count("*").alias("num_rides"), mean("days_between_rides").alias("mean_days_between_rides"))

# Compute top and bottom 10 riders:
rides4 \
  .where(rides4.mean_days_between_rides.isNotNull()) \
  .orderBy("mean_days_between_rides") \
  .show(10)

rides4 \
  .orderBy("mean_days_between_rides", ascending=False) \
  .show(10)

# **Question:** How can we make this analysis better?


# ## Exercises

# (1) What is the average time between rides for each driver?


# ## References

# [Spark Python API - pyspark.sql.Window
# class](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=catalog#pyspark.sql.Window)

# [Spark Python API - pyspark.sql.WindowSpec
# class](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=catalog#pyspark.sql.WindowSpec)


# ## Cleanup

# Stop the SparkSession:
spark.stop()

File: code-r/10_window.R
# # Window functions

# Copyright © 2010–2018 Cloudera. All rights reserved.
# Not to be reproduced or shared without prior written 
# consent from Cloudera.

# This module demonstrates how to use *window functions* with
# sparklyr. Window functions are a variation on aggregation 
# functions. Whereas an aggregation function, like `sum()`,
# groups sets of rows together into single rows, a window 
# function is applied over sets of rows *without* combining 
# them. Window functions are also called *analytic functions*.


# ## Setup

library(sparklyr)
library(dplyr)

config <- spark_config()
config$spark.driver.host <- Sys.getenv("CDSW_IP_ADDRESS")
spark <- spark_connect(
  master = "local",
  app_name = "window",
  config = config
)


# ## Load the rides data from HDFS into a Spark DataFrame

rides <- spark_read_parquet(
  sc = spark,
  name = "rides",
  path = "/duocar/clean/rides/"
)


# ## Window functions with dplyr

# Window functions in dplyr are implemented using the
# familiar dplyr verbs, but used in different
# combinations and in different order.

# Recall that to perform *standard* aggregation with
# dplyr, you use `summarise()`, preceded optionally by
# `group_by()` if you want to aggregate within groups.
# For example, to aggregate the total distance in miles
# of all DuoCar rides:

rides %>% 
  summarise(sum_distance_mi = sum(distance / 1609.34, na.rm = TRUE))

# The result shows that DuoCar drivers have provided 
# more than 250,000 miles of rides.

# We would like to identify which exact ride was the
# one during which the 250,000th mile was driven, so
# that we can give the driver and passenger a special
# award. To do this, we need to use window functions.

# Arrange the data by date and time, and calculate 
# the cumulative sum of distance (in miles):

cum_sum_miles <- rides %>% 
  arrange(date_time) %>%
  mutate(cum_sum_distance_mi = cumsum(distance / 1609.34))

# Let's limit the columns and look at the first few
# rows of the result DataFrame:

cum_sum_miles %>%
	arrange(date_time) %>%
	select(date_time, cum_sum_distance_mi) %>%
	head()

# and the last few rows of the result DataFrame:

cum_sum_miles %>%
	arrange(desc(date_time)) %>%
	select(date_time, cum_sum_distance_mi) %>%
	head()

# That looks right. To see the SQL that generates
# these results, use `show_query()`:

cum_sum_miles %>% show_query()

# Notice the `OVER` clause in the SQL. In SQL, the
# `OVER` clause defines a *window specification* which
# controls which rows the preceding function operates
# on.

# To find the ride during which the 250,000th mile
# was driven, filter to the rides with cumulative
# sum 250,000 or greater, arrange in ascending order
# by cumulative sum, and take the top row:

award_winner <- cum_sum_miles %>%
	filter(cum_sum_distance_mi >= 250000) %>%
	arrange(cum_sum_distance_mi) %>%
	head(1)

# Let's see which ride it is:

award_winner %>%
	select(id, rider_id, driver_id, distance, cum_sum_distance_mi)

# You could look up this ride ID in the joined data
# to find the driver's and rider's names and other
# details.

# The above example demonstrates how to compute a
# cumulative sum using `cumsum()`. You can also
# compute cumulative means using `cummean()`.

# Other applications of window functions include:
# * Calculating differences between group aggregates
#   and individual row values
# * Computing ranks (see `?ranking` for details)
# * Comparing values ahead of or behind the current
#   value (see `?lead-lag` for details)


# ## Exercise

# The joined data (in `/duocar/joined`) contains the columns
# `driver_rides` and `driver_stars` from the drivers table.
# These columns represent the total number of rides and stars
# that the driver has accumulated as of the time when the 
# data was generated. But these values are potentially 
# misleading, because these values do not represent the 
# values at the time when each ride occurred.
# Beginning with the joined data, compute a new dataset 
# with the `driver_rides` and `driver_stars` columns
# replaced by cumulative sums of the number of rides for each 
# driver and the `star_rating` column for each driver. These
# represent the actual values at the time each ride occurred.

# Confirm that the rides and stars values in the most recent
# entry for each driver in your result match the values in
# the existing joined data.


# ## Cleanup

spark_disconnect(spark)

# ## References

# [dplyr window functions vignette](https://cran.r-project.org/web/packages/dplyr/vignettes/window-functions.html)
